filebeat/generator/module/module.go

```diff
@@ -0,0 +1,33 @@
+package module
+
+import (
+	"fmt"
+	"path"
+
+	"github.com/elastic/beats/filebeat/generator"
+)
+
+// Generate generates a new module.
+// If module exists, error is returned.
+func Generate(module, modulesPath, beatsPath string) error {
+	modulePath := path.Join(modulesPath, "module", module)
+	if generator.DirExists(modulePath) {
+		return fmt.Errorf("module already exists: %s", module)
+	}
+
+	err := generator.CreateDirectories(modulePath, []string{path.Join("_meta", "kibana", "6")})
+	if err != nil {
+		return err
+	}
+
+	replace := map[string]string{"module": module}
+	templatesPath := path.Join(beatsPath, "scripts", "module")
+	filesToCopy := []string{
```

Comment:
it calls `generator.CopyTemplates` which not only copies the files, but substitutes the template strings in copied files. I think apart from copying, there is not so much to share.

I could move `Copy` from the package `mage` to `libbeat/utils` or `libbeat/common/utils` and use it from there. It would be a bit bigger undertaking, as right now files are copied one-by-one and after each file is copied, the substitution is done. If I use this, I need to change the algorithm to substitute template strings after the file is copied. It would also mean that each file would need to be opened, read and written twice, once to copy it and once to substitute strings.

However, if I do so, I would rather refactor all generators and other possible places where files are copied. So I would rather address this in a follow up PR. WDYT?

Reply:
SGTM